File: 108. Convert Sorted Array to Binary Search Tree/108.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    """
        Title: 108. Convert Sorted Array to Binary Search Tree
        Time Complexity: O(N)
        Algorithm: 
            1. Divide & Conquer
        
        Note: 
            1. If the size of the left subtree and that of the right subtree
               are the same or at most different by one, for every node, then
               it's a perfect binary search tree(for the h-1 levels), also a 
               height-balanced BST
        
        Date: 07/17/2021 00:39		
        Time: 52 ms  -- beat 97.01%
        Memory: 15.7 MB  -- beat 17.24%
    """

    def dfs(self, i, j) -> TreeNode:
        if i == j:
            return TreeNode(self.nums[i])
        
        mid = (i + j) // 2
        root = TreeNode(self.nums[mid])
        
        if i < mid:
            root.left = self.dfs(i, mid-1)
        
        if mid < j:
            root.right = self.dfs(mid+1, j)
        
        return root
    
    def sortedArrayToBST(self, nums: List[int]) -> TreeNode:
        self.nums = nums
        return self.dfs(0, len(nums)-1)

File: tagged/sprint/binaryTree/108. Convert Sorted Array to Binary Search Tree/108.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

#include <vector>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    TreeNode* construct(vector<int>& nums, int left, int right) {
        if (left > right) {
            return nullptr;
        }
        
        int mid = (left + right) / 2;
        TreeNode* root = new TreeNode(nums[mid]);
        root->left = construct(nums, left, mid-1);
        root->right = construct(nums, mid+1, right);
        return root;
    }
    
    TreeNode* sortedArrayToBST(vector<int>& nums) {
        return construct(nums, 0, nums.size()-1);
    }
};